src/main/kotlin/org/rust/ide/inspections/checkMatch/Pattern.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections.checkMatch
+
+import org.rust.lang.core.psi.RsEnumVariant
+import org.rust.lang.core.psi.RsStructItem
+import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.psi.ext.RsFieldsOwner
+import org.rust.lang.core.psi.ext.findInScope
+import org.rust.lang.core.psi.ext.parentEnum
+import org.rust.lang.core.types.ty.Ty
+import org.rust.lang.core.types.ty.TyAdt
+import org.rust.lang.core.types.ty.TyTuple
+import org.rust.lang.core.types.ty.TyUnknown
+
+data class Pattern(val ty: Ty, val kind: PatternKind) {
+    fun text(ctx: RsElement? = null): String =
+        when (kind) {
+            is PatternKind.Wild -> "_"
+
+            is PatternKind.Binding -> kind.name
+
+            is PatternKind.Variant -> {
+                val variantName = kind.variant.name.orEmpty()
+                val itemName = kind.item.name
+
+                val variantInScope = ctx?.findInScope(variantName)
+                    ?.takeIf { (it as? RsEnumVariant)?.parentEnum == kind.item }
+
+                // if the variant is already in scope, it can be used as just `A` instead of full `MyEnum::A`
+                val name = if (variantInScope != null) {
+                    variantName
+                } else {
+                    "$itemName::$variantName"
+                }
+                val initializer = kind.variant.initializer(kind.subPatterns)
+                "$name$initializer"
+            }
+
+            is PatternKind.Leaf -> {
+                val subPatterns = kind.subPatterns
+                when (ty) {
+                    is TyTuple -> subPatterns.joinToString(", ", "(", ")") { it.text() }
+                    is TyAdt -> {
+                        val name = (ty.item as? RsStructItem)?.name ?: (ty.item as? RsEnumVariant)?.name
+                        val initializer = (ty.item as RsFieldsOwner).initializer(subPatterns)
+                        "$name$initializer"
+                    }
+                    else -> ""
+                }
+            }
+
+            is PatternKind.Range -> "${kind.lc}${if (kind.isInclusive) ".." else "..="}${kind.rc}"
+
+            is PatternKind.Deref -> "&${kind.subPattern.text()}"
```

Comment:
Looks like we should propagate context here.
The quickfix still produces `Option::Some`  on this example:
```rust
match &Some(1) {
    &None => {},
}
```
![Peek 2019-03-09 11-23](https://user-images.githubusercontent.com/3221931/54068460-fd441500-425d-11e9-828a-7aac1aed8b72.gif)

Reply:
Yep, now it propagates in all cases